### DISCUSSION TITLE
What about security?

### DISCUSSION THREAD

**Discussion by georgestevens99** [adamreed90/dapr-samples#7] [Ideas]
In most of my Dapr POCs to date I have just ignored security and had none.  I use port 80 and NO certificate, NO developer secrets, and only hardcoded passwords in the yaml for Redis, etc.  These POCs are totally hackable, but they seldom run, they have no sensitive data, and they avoid a significant amount of work in generating certificates, etc to setup for good security.  However I have done a few whose main focus was on security, rather than other functionality.

I am open to do good security for this if the contributors want to do the work required to secure it.  However, doing that means the people using the sample will have to do the security implementation and we will have to write something to tell them how to do that.

So, since the main goals are 1) Demo Service Invocation in .NET and 2) Have a set of services that can be reused for various other samples in the future, I suggest we save security for later samples.

What do you guys think? @adamreed90 @paulgyro 

Thanks!

**Comment by iamadamreed** [comment]
I definitely think it's important to show using the security store and doing it properly. A true sample, not a basic tutorial is my personal goal. I concur! Maybe we do one of the services unsecured for tutorials sake however, but make it clear.

**Reply by paulgyro** [reply]
Maybe we should have a progression of solutions, one without the security and a final one that adds all the security elements?  Ether, we do as I mentioned OR we start with "security first" which is AWS mantra.

Thoughts?

**Reply by iamadamreed** [reply]
I agree, I think it's important to illustrate both, as for example in my daily work environment, we have an entirely air gapped service that doesn't require excessive security, as there's no access possible, but we definitely want a "by the book" example as well.

**Reply by georgestevens99** [reply]
OK!  Lets try this out with Service Invocation, and if it works then we can use it in samples of other Dapr capabilities.

So our key Customer Goals for this repo are:
Produce code samples that:
1.  Newbies can use to quickly get running to understand how to think of and use Dapr capabilities in their code.
2.  Any level of developer can use to understand deeper aspects of Dapr capabilities, like Security, advanced usage scenarios, etc.

To that end, how about if we develop a set of stand alone samples, all with a similar basic code structure like we have in our repo now i.e.  console --> WebAPIA --> grpcServiceA.  This will allow us to to have a "progression" of solutions, although we will most likely have to tolerate code duplication which is OK since we shouldn't have to do much maintenance.

Here is how we could do that.  Use subfolders in dapr-samples/hello-dotnet6, shown below:
dapr-samples/hello-dotnet6/service-invocation-basic
dapr-samples/hello-dotnet6/service-invocation-with-security
dapr-samples/hello-dotnet6/statemanagement-basic
dapr-samples/hello-dotnet6/statemanagement-with-security
dapr-samples/hello-dotnet6/pubsub-basic
dapr-samples/hello-dotnet6/pubsub-with-security
etc.

So all we have to do now is get the first one operational, dapr-samples/hello-dotnet6/service-invocation-basic, and then we can use that for the basis of the next one.   We evolve things as we go, and I feel it is not necessary to keep backfilling any improvements we make to newer folders to the existing code.

Keep in mind that 1) We will need to write a document for dapr/samples/hello-dotnet6 and 2) make sure the above scheme is OK with Dapr management.

So will the above proposal keep everyone satisfied, while allowing us to get moving quickly?  It will only require minimal changes to what is there now.
@paulgyro

**Reply by iamadamreed** [reply]
That sounds good to me!

**Reply by georgestevens99** [reply]
PS  When all parties agree to this evolutionary path, or some revision thereof, I will snip out a copy of the text that describes it and submit the description as part of a proposal to the dapr/samples repo through a new issue.  I will ask for their feedback and approval of our proposal so that everyone is on the same page and we have assurance that our work will not be sidetracked.  PS.  It may be a good idea for each of us to review the dapr/samples contribution guidelines or what ever they call it.
